mongo/datadog_checks/mongo/mongo.py

```diff
@@ -159,36 +162,31 @@ def _build_metric_list_to_collect(self):
         return metrics_to_collect
 
     def _refresh_replica_role(self):
-        if self._api_client and (
-            self._api_client.deployment_type is None
-            or isinstance(self._api_client.deployment_type, ReplicaSetDeployment)
-        ):
+        if self.api_client.deployment_type is None or isinstance(self.api_client.deployment_type, ReplicaSetDeployment):
             self.log.debug("Refreshing deployment type")
-            self._api_client.deployment_type = self._api_client.get_deployment_type()
+            self.api_client.refresh_deployment_type()
 
     def check(self, _):
-        if self._connect():
-            self._check()
+        try:
+            self._refresh_metadata()
+            self._collect_metrics()
+            self.service_check(SERVICE_CHECK_NAME, AgentCheck.OK, tags=self._config.service_check_tags)
+        except CRITICAL_FAILURE:
+            self.service_check(SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self._config.service_check_tags)
+            self._unset_metadata()
```

Comment:
Could we have some other kind of exceptions here that would not be caught?

Reply:
The current implementation of `_unset_metadata` just logs and sets a variable to `None`, so I'm pretty confident it won't be a problem. I think if we ever change that implementation we'll have to revisit this place for sure.

Reply:
😅  sorry I was not clear enough. I mean: Could we get another kind of exception besides the ones in `CRITICAL_FAILURE`, in which case we would not send the service check (but maybe that's what we want in that case?)

Reply:
The current `CRITICAL_FAILURE` members are based on what we test for explicitly and what I read in [the docs](https://pymongo.readthedocs.io/en/latest/api/pymongo/errors.html).

I feel least confident about this part of the change though, and would not be surprised if we'd have to extend this list based on support cases :(

Reply:
I've added a couple more members based on the pymongo docs